csp/impl/types/pydantic_type_resolver.py

```diff
@@ -0,0 +1,209 @@
+import numpy
+from pydantic import TypeAdapter, ValidationError
+from typing import Any, Dict, List, Set, Tuple, Type, Union, get_args
+
+import csp.typing
+from csp.impl.types.container_type_normalizer import ContainerTypeNormalizer
+from csp.impl.types.instantiation_type_resolver import UpcastRegistry
+from csp.impl.types.numpy_type_util import map_numpy_dtype_to_python_type
+from csp.impl.types.pydantic_types import CspTypeVarType, adjust_annotations
+from csp.impl.types.typing_utils import CspTypingUtils, TsTypeValidator
+
+
+class TVarValidationContext:
+    """Custom validation context class for handling the special csp TVAR logic."""
+
+    # Note: some of the implementation is borrowed from InputInstanceTypeResolver
+
+    def __init__(
+        self,
+        forced_tvars: Union[Dict[str, Type], None] = None,
+        allow_none_ts: bool = False,
+    ):
+        # Can be set by a field validator to help track the source field of the different tvar refs
+        self.field_name = None
+        self._allow_none_ts = allow_none_ts
+        self._forced_tvars: Dict[str, Type] = forced_tvars or {}
+        self._tvar_type_refs: Dict[str, Set[Tuple[str, Type]]] = {}
+        self._tvar_refs: Dict[str, Dict[str, List[Any]]] = {}
+        self._tvars: Dict[str, Type] = {}
+        self._conflicting_tvar_types = {}
+
+        if self._forced_tvars:
+            config = {"arbitrary_types_allowed": True}
+            self._forced_tvars = {k: ContainerTypeNormalizer.normalize_type(v) for k, v in self._forced_tvars.items()}
+            self._forced_tvar_adpaters = {
+                tvar: TypeAdapter(List[t], config=config) for tvar, t in self._forced_tvars.items()
```

Comment:
It's a List[t] because we accumulate multiple values that were assigned to the type var and try to convert them all at once (i.e. as a List[t]) as it's more efficient than trying to convert them individually.